packages/core/src/input/InputManager.ts

```diff
@@ -0,0 +1,320 @@
+    if (evtTypeList <= 1 && _activePointerCount === 0) {
+      return;
+    } else {
+      for (let i = 0; i < _activePointerCount; i++) {
+        pointers[i] = pointerList[i];
+      }
+    }
+
+    /** Check whether pressed events are triggered. */
+    let prePressedEntity = input.pressedEntity;
+    if (prePressedEntity) {
+      const preScripts = prePressedEntity._scripts._elements;
+      for (let i = preScripts.length - 1; i >= 0; i--) {
+        preScripts[i].onPointerDrag();
+      }
+    }
+    /** Get the entity hit by the ray. */
+    const curEntity = this._pointerRaycast(input.x, input.y);
+    const preEnteredEntity = input.enteredEntity;
+    /** 80% of operations are on curEntity, so we cache his scripts. */
+    let curEntityScripts: Script[];
+    let curEntityScriptsLen = 0;
+    if (curEntity) {
+      curEntityScripts = curEntity._scripts._elements;
+      curEntityScriptsLen = curEntityScripts.length;
+    }
+    /** Check whether enter and exit events are triggered. */
+    if (preEnteredEntity != curEntity) {
+      if (curEntity) {
+        for (let i = 0; i < curEntityScriptsLen; i++) {
+          curEntityScripts[i].onPointerEnter();
+        }
+      }
+      if (preEnteredEntity) {
+        const scripts = preEnteredEntity._scripts;
+        const scriptsArr = scripts._elements;
+        for (let i = scripts.length - 1; i >= 0; i--) {
+          scriptsArr[i].onPointerExit();
+        }
+      }
+      input.enteredEntity = curEntity;
+    }
+    /** Check whether down up and click events are triggered. */
+    while (evtTypeList > 1) {
+      switch (evtTypeList % 2) {
+        case PointerEventType.Down:
+          if (curEntity) {
+            for (let j = 0; j < curEntityScriptsLen; j++) {
+              curEntityScripts[j].onPointerDown();
+            }
+          }
+          prePressedEntity = curEntity;
+          break;
+        case PointerEventType.Up:
+          if (prePressedEntity) {
+            if (prePressedEntity == curEntity) {
+              for (let j = 0; j < curEntityScriptsLen; j++) {
+                const script = curEntityScripts[j];
+                script.onPointerUp();
+                script.onPointerClick();
+              }
+            } else {
+              const preScripts = prePressedEntity._scripts._elements;
+              for (let j = preScripts.length - 1; j >= 0; j--) {
+                preScripts[j].onPointerUp();
+              }
+            }
+          }
+          prePressedEntity = null;
+          break;
+      }
+      evtTypeList >>= 1;
+    }
+    input.pressedEntity = prePressedEntity;
+  }
+
+  /**
+   * Get the Entity to which the ray is cast.
+   * @param x - The X coordinate of the pointer on the screen, specified in normalized
+   * @param y - The Y coordinate of the pointer on the screen, specified in normalized
+   * @returns The Entity to which the ray is cast
+   */
+  private _pointerRaycast(x: number, y: number): Entity {
```

Comment:
The `input.x and input.y ` maybe  should canvas pixel unit and call `camera.screenPointToRay()` directly.

Reply:
Done.